src/dynarec/rv64/dynarec_rv64_f0.c

```diff
@@ -339,6 +339,59 @@ uintptr_t dynarec64_F0(dynarec_rv64_t* dyn, uintptr_t addr, uintptr_t ip, int ni
                     DEFAULT;
             }
             break;
+        case 0x10:
+            INST_NAME("LOCK ADC Eb, Gb");
+            READFLAGS(X_CF);
+            SETFLAGS(X_ALL, SF_SET_PENDING);
+            nextop = F8;
+            GETGB(x2);
+            SMDMB();
+            if((nextop&0xC0)==0xC0) {
+                if(rex.rex) {
+                    wback = xRAX + (nextop&7) + (rex.b<<3);
+                    wb2 = 0;
+                } else {
+                    wback = (nextop&7);
+                    wb2 = (wback>>2);
+                    wback = xRAX+(wback&3);
+                }
+                SRLIW(x1, wback, wb2*8);
+                ANDI(x1, x1, 0xFF);
+                emit_adc8(dyn, ninst, x1, x2, x3, x4, x5);
+                SLLI(x1, x1, wb2*8);
+                MOV_U12(x3, 0xFF);
+                SLLI(x3, x3, wb2*8);
+                NOT(x3, x3);
+                AND(wback, wback, x3);
+                OR(wback, wback, x1);
+            } else {
+                addr = geted(dyn, addr, ninst, nextop, &wback, x1, x3, &fixedaddress, rex, LOCK_LOCK, 0, 0);
+                LBU(x3, wback, 0);
+                emit_adc8(dyn, ninst, x3, x2, x4, x5, x6);
+                SB(x3, wback, 0);
+            }
+            SMDMB();
+            break;
+        case 0x11:
+            INST_NAME("LOCK ADC Ed, Gd");
+            READFLAGS(X_CF);
+            SETFLAGS(X_ALL, SF_SET_PENDING);
+            nextop = F8;
+            GETGD;
+            SMDMB();
+            if(MODREG) {
+                ed = xRAX+(nextop&7)+(rex.b<<3);
+                emit_adc32(dyn, ninst, rex, ed, gd, x3, x4, x5, x6);
+            } else {
+                addr = geted(dyn, addr, ninst, nextop, &wback, x2, x1, &fixedaddress, rex, LOCK_LOCK, 0, 0);
+                MARKLOCK;
+                LRxw(x1, wback, 1, 1);
+                emit_adc32(dyn, ninst, rex, x1, gd, x3, x4, x5, x6);
```

Comment:
I don't think this will work if there are some flags computations. There will be too many instruction between the read and the write.
You should do the ADC inline between the read/write (with precomputing the Carry before the read) and preserving the initial regs if needed, and than call adc32 if there are some flags needed outside the read/write sequence.

Reply:
A better way might be to use the atomic add, with pre-adding the Carry to GD before.